### ISSUE TITLE
Configurable perf metrics

### ISSUE THREAD

**Issue by tilsche** [tud-zih-energy/lo2s#8]
- [x] Select perf metrics through command line
- [x] Add program_option, which prints all available metrics
- [ ] Sample perf metrics also for -a
- [x] Configurable interrupt source
- [x] Use `PERF_FORMAT_GROUP` and reduce number of reads

**Comment by tilsche** [comment]
Regarding interrupt source:
Rename `--sampling-period` to `--count` for perf record compatibility. `-e,--event=...` is not the interrupt event -> save in config -> perf. `-E,--metric-event` is the old `-e`.

**Comment by tilsche** [comment]
1. Implement a `PerfCounterGroup`
2. Use `PERF_SAMPLE_READ` to put metric reads in a ringbuffer. Select appropriate group leader (unhalted reference cycles) `PERF_SAMPLE_PERIOD`.
3. Change `ThreadMonitor` from `IntervalMonitor` to `FdMonitor`. Monitors the sample fd for a thread and the PerfCounter fd
4. Deal with perf metrics in `-a`

For the future, we will make a runtime configurable monitor baseclass that selects either the poll/FdMonitor or IntervalMonitor main loop.

**Comment by cvonelm** [comment]
@phijor could you give me a few pointers towards what you've previously done on this issue?

**Comment by bmario** [comment]
There was this PR: #26 
Do you have any specific questions?

**Comment by cvonelm** [comment]
nothing at the moment, this seems as a good entrance to the issue

**Comment by cvonelm** [comment]
The issue-8 branch now contains a first implementation of metrics in system mode(Screenshot below, which shows some normal activity from my computer with the 'instructions' metric enabled). It would be nice if someone, with more knowledge of how a 'correct' trace should look, could take a look at it and tell me if the output this branch is producing is reasonable
![screenshot from 2018-04-30 14-45-24](https://user-images.githubusercontent.com/17126443/39427737-30553ee4-4c85-11e8-9cf5-2ff61d65eceb.png)

**Comment by cvonelm** [comment]
As requested, I've made some measurements using the code from the issue-8 branch in system and normal mode.
I've used this  command line
```bash
./lo2s [-a] -e instructions env OMP_NUM_THREADS=32 OMP_PLACES=cores OMP_PROC_BIND=true /home/tilsche/npb/NPB3.3.1/NPB3.3-OMP/bin/bt.C.x
```
As GOMP_CPU_AFFINITY is GNU specific and the NPB suite used here was compiled with Intel Fortran I've used the options above instead, which basically end up in the same result.

As far as I have seen it the values both traces have recorded are very similiar. Of course the exact course is different, but as far as Average, Maximum, Minimum Values, and the general form of the graph are concerned they are very similar. It seems that the measurement mode works as expected

As far as performance goes, there is no noticable difference in the self reported performance indicators of the NPB suite for both measurement modes

![similarities](https://user-images.githubusercontent.com/17126443/40365009-e7fdeae6-5dd3-11e8-8883-8a7bb7fe944f.png)
